Add packet fragmentation

TimonPost opened the pull request:
This is an implementation of fragmentation ([issue](https://github.com/LucioFranco/amethyst-protocol/issues/5)). Looks like it is working, but need to test it more still. So work in progress...

Note that I have looked at this [libary](https://github.com/jaynus/netcode.io/tree/master/rust) which had some basic implementation of fragmentation. It is from the author of gafferongames blogs: 

**Things I changed:**

1. There are two headers:

- PacketHeader
This will contain all the header stuff we first had in  RawPacket. This will be used for every packet.

- Fragment Header
Fragment header the information of the fragments a packet is divided into. The first fragment packet will have two headers. The standard PacketHeader and the Fragment header the following fragments only have the fragment header. 

2. Also, I have changed to using `byteorder` so that I can control more better how a header gets serialized an deserialized.
3. I have a FragmentBuffer which will be used to store fragments temporarily
4. Added a config ty[e for network settings.

**Still, need to do:**
1. Tests
2. Config file needs to be given by the user (for now the default is used).
3. I want to make the preproccess_packet function smaller it is pritty big now.

LucioFranco commented on src/net/udp.rs:
```diff
@@ -1,48 +1,61 @@
         let socket = net::UdpSocket::bind(addr)?;
         let state = SocketState::new();
+        // TODO: add functionality to get config from file.
+        let config = NetworkConfig::default();
```
I think we should just pass this in, let the user decide how to create the config.

LucioFranco commented on src/net/udp.rs:
```diff
@@ -1,48 +1,61 @@
+    pub fn send(&mut self, mut packet: Packet) -> Result<usize> {
+        let (addr, mut packet_data) = self.state.pre_process_packet(packet, &self.config)?;
+
+        let mut bytes_send = 0;
```
`bytes_sent`

LucioFranco commented on src/packet/header/fragment.rs:
```diff
@@ -0,0 +1,128 @@
+                    wtr.write(&header.parse()?)?;
+                },
+                None => {
+                    return Err(Error::new(ErrorKind::Other, "Invalid fragment header"));
```
Should have an error type for this

TimonPost replied:
There is one problem of doing that you have to see the context of the whole function. 

Every 'write' in that function returns an `io::Result`. So we could paste this line `.map_err(|| NetworkError::InvalidPacketHeader.into()?);` behind each write function or we could simply return `io::Result` and let the caller of that function maps the error to `.map_err(|| NetworkError::InvalidPacketHeader.into()?);` 

I think that is better otherwise this function will look horrible.

LucioFranco replied:
In reality we should have a Error::from implemented for `io::Result`

LucioFranco commented on src/packet/header/packet.rs:
```diff
@@ -0,0 +1,85 @@
+    /// Get the size of this header.
+    pub fn size(&self) -> u8
+    {
+        return PACKET_HEADER_SIZE;
```
You dont need the return keyword here 😄

LucioFranco commented on src/packet/packet_processor.rs:
```diff
@@ -0,0 +1,220 @@
+        }
+
+        return match received_bytes {
+            Ok(Some(payload)) => return Ok(Some(Packet::new(addr, payload))),
```
drop the return here

LucioFranco commented on src/packet/packet_processor.rs:
```diff
@@ -0,0 +1,220 @@
+
+        return match received_bytes {
+            Ok(Some(payload)) => return Ok(Some(Packet::new(addr, payload))),
+            Ok(None) => return Ok(None),
```
Same as above

TimonPost commented on src/net/socket_state.rs:
```diff
@@ -40,30 +42,63 @@ impl SocketState {
-        Ok((packet.addr, buffer))
+        let connection_seq = lock.seq_num;
+        // queue new packet
+        lock.waiting_packets.enqueue(connection_seq, packet.clone());
```
For now, it does not really matter much though because the packet is not changed after it is enqueued waiting for acknowledgment. The packet is content will be ready before it is enqueued. What we can do to avoid change is to make the fields of Packet nonpublic and only readable. So once a Packet is defined you can change it anymore.

TimonPost replied:
I have changed all fields of the packet to private and made getters for them so they can not be changed once declared.

LucioFranco replied:
Okay, thats fine, I was just thinking that this would be a good use case for a ref counter.

TimonPost commented on src/net/socket_state.rs:
```diff
@@ -40,30 +42,63 @@ impl SocketState {
+                let fragment = FragmentHeader::new(fragment_id, num_fragments, packet_header.clone());
+
+                // get start end pos in buffer
+                let start_fragment_pos = fragment_id as u16 * config.fragment_size; /* upcast is safe */
```
We do store it the fragment_size as u16. I think you mean fragment_id? But that is an u8 so the upcast is safe here. To multiply fragment_id with fragment size they need to be of the same type. And by casting the fragment_id to u16 we can multiply those two numbers.

TimonPost commented on src/error.rs:
```diff
@@ -16,5 +16,21 @@ pub enum NetworkError {
+    HeaderParsingFailed,
+    #[fail(display = "Something went wrong when sending")]
+    SendFailed,
+    #[fail(display = "Something went wrong when receiving")]
```
we will do this later

TimonPost commented on src/error.rs:
```diff
@@ -16,5 +16,21 @@ pub enum NetworkError {
+    ExceededMaxFragments,
+    #[fail(display = "The packet header is invalid.")]
+    InvalidPacketHeader,
+    #[fail(display = "Just an error where we don't care about the type.")]
```
Yea can be removed had it for test forgot to remove it.

TimonPost commented on Cargo.toml:
```diff
@@ -10,6 +10,7 @@ license = "MIT/Apache-2.0"
 serde = "1.0"
 serde_derive = "1.0"
 bincode = "1.0"
+byteorder = "*"
```
no, no reason going to set it to an version good point!

LucioFranco replied:
Agreed.

TimonPost commented on src/lib.rs:
```diff
@@ -18,4 +18,39 @@ pub mod error;
+/// - number 1 / number 2 = 3,90625 (this will be rounded to 3)
+/// - add remainder from above to 3 = 4.
+///
+/// An you will that for all numbers you push into this method it will return an nice result.
```
Gotcha must be copy paste work :)

TimonPost commented on src/net/socket_state.rs:
```diff
@@ -40,30 +42,63 @@ impl SocketState {
-        let mut l = connection
+        let connection = self.create_connection_if_not_exists(&packet.addr)?;
+
+        let mut lock = connection
```
fixed!

TimonPost commented on src/net/constants.rs:
```diff
@@ -0,0 +1,3 @@
+pub const DEFAULT_MTU: usize = 1452;
```
Jup, further investigation needed. I'll keep it like this for now. If you wondered why 1452 (gaffer recommended size): 

/// Derived from ethernet_mtu - ipv6_header_size - udp_header_size - packet header size
///       1452 = 1500         - 40               - 8               - 8

TimonPost commented on src/packet/fragment_buffer.rs:
```diff
@@ -0,0 +1,109 @@
+        let index = self.index(sequence);
+
+        if self.entry_sequences[index] != sequence {
+            return None;
```
We need to return here other wise I'll get the error expected `()` found Option

fhaynes commented on src/error.rs:
```diff
@@ -1,9 +1,17 @@
 pub type Error = failure::Error;
 pub type Result<T> = result::Result<T, Error>;
 
+//impl From<io::Error> for Error
```
Can we not leave commented code lying about? I know we're going to redo error-handling, but I think its good habit in general to not leave commented out code in the code, and instead put it in an issue

torkleyy commented on src/net/network_config.rs:
```diff
@@ -1,21 +1,39 @@
     pub max_packet_size: usize,
+    /// These are the maximal fragments a packet could be divided into.
+    ///
+    /// Why can't I have more than 256 (u8)?
```
super minor nit: 255

torkleyy commented on src/net/network_config.rs:
```diff
@@ -1,21 +1,39 @@
+    /// These are the maximal fragments a packet could be divided into.
+    ///
+    /// Why can't I have more than 256 (u8)?
+    /// This is because you don't want to send more then 256 fragments over UDP, with high amounts of fragments the change for an invalid packet is very high.
```
"change" -> "chance"

Dumb question from somebody who doesn't know much about networking: If one packet failed, couldn't we just resend it? I think it would be ideal if the size of the packet can be as large as necessary, so the user does not have to split very big messages manually.

TimonPost replied:
One time stuff? What we want to have like TCP: ordered +  able to see dropped packets (not automatically resent them) + the fragmentation (this PR, and later to be able to send each fragment individually if dropped).

fhaynes replied:
@torkleyy You can also think of it this way: TCP is built on top of UDP. We're doing the same thing, but only implementing the few features of TCP that are useful to a network game.

Regarding packet size, when sending data over the Internet, in general you want to keep the packet size at less than 1500 bytes. The reason for this is historical. Thus, we have to implement our own fragmentation, since plain UDP does not have it.

torkleyy replied:
@TimonPost I meant things like assets. I'm just reading all the documents and @fhaynes answers my questions on Discord ;) I understood this now, you can see this as resolved except for "change" which should be "chance".

TimonPost replied:
I'll have already changed that thanks for looking into the PR

torkleyy commented on src/lib.rs:
```diff
@@ -3,19 +3,54 @@
+/// But since we care about how how many whole times the number fits in we need the result 4.
+///
+/// Note that when rust is rounding it is always rounding to zero (3.456 as u32 = 3)
+/// 1. calculate with modules if `number 1` fits exactly in the `number 2`.
```
not sure if you meant "modulo"?

torkleyy commented on src/lib.rs:
```diff
@@ -3,19 +3,54 @@
+/// [fragment: 1024] [fragment: 1024] [fragment: 1024] [fragment: 928]
+fn total_fragments_needed(payload_length: u16, fragment_size: u16) -> u16
+{
+    let remainder = if payload_length % fragment_size > 0 { 1 } else { 0 };
```
How about `max((payload_length + fragment_size - 1) / fragment_size, 1)`?

TimonPost replied:
I come back at this need to take a look if this works.

torkleyy commented on src/packet/header/fragment.rs:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    /// Get the size of this header.
+    pub fn size(&self) -> u8
```
These methods should probably go into a trait.

TimonPost replied:
Good point going to do that.

LucioFranco commented on src/error.rs:
```diff
@@ -1,9 +1,13 @@
 pub type Error = failure::Error;
 pub type Result<T> = result::Result<T, Error>;
 
+
```
Remove these new lines please

LucioFranco commented on src/packet/fragment_buffer.rs:
```diff
@@ -0,0 +1,109 @@
+use std::io::Result;
+use std::clone::Clone;
+
+/// Collection for storing data fof any kind.
```
`fof` should be `of`

LucioFranco commented on src/packet/fragment_buffer.rs:
```diff
@@ -0,0 +1,109 @@
+
+    /// Remove entry from collection.
+    pub fn remove(&mut self, sequence: u16) {
+        // TODO: validity check
```
Could you explain more what you mean by this `TODO`.

LucioFranco commented:
@fhaynes btw
````
error: expected `;`, found `use`
 --> src/net/udp.rs:5:1
  |
4 | use super::{constants, Packet, RawPacket}
  |                                          - expected `;`
5 | use socket_state::SocketState; 
  | ^^^ unexpected token
```

fhaynes commented:
> @fhaynes btw
> 
> ```
> error: expected `;`, found `use`
>  --> src/net/udp.rs:5:1
>   |
> 4 | use super::{constants, Packet, RawPacket}
>   |                                          - expected `;`
> 5 | use socket_state::SocketState; 
>   | ^^^ unexpected token
> ```
> ```

Fixed, thanks.

TimonPost commented:
Oke fixed the merge, ran all the tests with success, removed some unused imports and fixed all your comments. So going to merge now since you both have approved.
